File: 26_dp_ii/07_palindrome_partitioning.rs
use std::{cmp::min, collections::HashMap};

fn palindrom_partition(s: String) -> usize{
    let mut dp: HashMap<usize, usize> = HashMap::new();
    return helper(&s, 0, &mut dp)
}

fn helper(s: &str, ind: usize, dp: &mut HashMap<usize, usize>) -> usize{
    if dp.contains_key(&ind){
        return *dp.get(&ind).unwrap()
    }
    if &s[ind..] == &s[ind..].chars().rev().collect::<String>(){
        return 0
    }
    let mut res = s.len();
    
    for i in ind+1..s.len(){
        let curr_string = &s[ind..i];
        if curr_string == curr_string.chars().rev().collect::<String>(){
            res = min(res, 1 + helper(s, i, dp))
        }
    }
    dp.insert(ind, res);
    res
}


fn main(){
    assert_eq!(palindrom_partition("ababbbabbababa".to_string()), 3);
    assert_eq!(palindrom_partition("aaabba".to_string()), 1);
    assert_eq!(palindrom_partition("aaa".to_string()), 0);
}

File: 26_dp_ii/07_palindrome_partitioning.go
package main

import "fmt"

func palindrom_partitioning(s string) int {
	dp := map[int]int{}
	var helper func(ind int) int

	helper = func(ind int) int {
		if val, ok := dp[ind]; ok {
			return val
		}
		if is_palindrom(s[ind:]) {
			return 0
		}
		res := len(s)
		for i := ind + 1; i < len(s); i++ {
			curr := s[ind:i]
			if is_palindrom(curr) {
				res = min(res, 1+helper(i))
			}
		}
		dp[ind] = res
		return res
	}
	return helper(0)
}

func is_palindrom(s string) bool {
	i, n := 0, len(s)
	for i <= n/2 {
		if s[i] != s[n-i-1] {
			return false
		}
		i += 1
	}
	return true
}

func main() {
	fmt.Println(palindrom_partitioning("ababbbabbababa") == 3)
	fmt.Println(palindrom_partitioning("aaabba") == 1)
	fmt.Println(palindrom_partitioning("aaa") == 0)
}
